### ISSUE TITLE
When the emails are deleted, frappe.model.delete_doc.delete_dynamic_links causes 100% CPU load

### ISSUE THREAD

**Issue by joolsr** [frappe/erpnext#8934]
**Problem**

Using ERPNext: v8.0.29 (master)
Frappe Framework: v8.0.47 (master)

Since the commit below

https://github.com/frappe/frappe/commit/ebbcd4593e3cb7fac06f2b3ad9c05fc3f4cf88dc

emails in the email inbox view when deleted (not moved to trash) , calls **frappe.model.delete_doc.delete_dynamic_links** which runs for approx 30s per deleted email and causes 100 % CPU load.

My server is not under specced. For 5 users (just in test situation only as yet), it is a dual core, SSD with 4GB RAM installed with Ubuntu 16.04 LTS with the easy-install script as a vanilla install.

It appears that the background task when it calls the process to delete dynamic links is at fault somewhere which was refactored with the commit above. Before this commit emails when deleted, and there was no slowing of the box that was noticeable.

This a forum post I made that gives a bit more background.
https://discuss.erpnext.com/t/getting-100-cpu-when-emails-and-communications-are-deleted/23956

If you delete 10 emails, then it does take ten times longer to for the high cpu activity to cease. ie for ten emails about 3-4 minutes on average. 

I'm not a programmer but happy to help provide info to help with the debug

**Comment by joolsr** [comment]
I should also add that thsi of course impacts other background jobs ie sendmail, search update. So there are many issues that arise beacsue of the blocking by the dynamic link process
ie
![backgroundjobsqueue](https://cloud.githubusercontent.com/assets/934876/26249708/810b1b52-3c9f-11e7-9a0b-111e5c11d067.png)

**Comment by tundebabzy** [comment]
hello @joolsr 
Thanks for reporting this. If you are still experiencing this, let's start by finding out what exactly is causing your cpu to spike that much. On your test setup, replicate the problem again and then at the same time, fire up your terminal and run the `top` command. This will show you all the processes your CPU is running and their memory utilization. Please report back with the culprit or post a screenshot here

**Comment by joolsr** [comment]
hello@tundebabzy
Please see the gif below. It doesn't say too much just mysql processes heavily loading the server. Pls note apart from deleting emails the box rarely gets loaded much at all.

![peek 2017-05-26 18-19](https://cloud.githubusercontent.com/assets/934876/26505253/f2667bfc-423f-11e7-979a-a006031aceb1.gif)

I guess we need to debug further I'm happy to run some other mysql stuff when the problem happens,

**Comment by tundebabzy** [comment]
@joolsr 
Can you assist us by installing a GUI to monitor the database like dbForge or MySQL Workbench?

**Comment by joolsr** [comment]
Hello again

As a test, I tested my gmail account on my test account on erpnext.com ie joolsr.erpnext.com. When I deleted many emails ie 15 or more there is no loading or importantly background tasks with where its working with dynamic links as above and hence no mysql processes for more than only a few seconds at most.

So something in an earlier commit  has caused problems for me. 

Oddly enough I also noticed that the css that is shown inline in many marketing emaisl also completely disappears in this instance. I have seen forum posts where peopel have complained about the same issues as me with seeing the css text so I think somehow there are issues relating to updates compared to a fresh install but I;'m not sure of this.

I also created a fresh install on a clean VPS with same spec as before. Again emails can be deleted without issues, css rendered perfectly and also issues I got with 'Communication not found '  (which I raised as a #https://github.com/frappe/erpnext/issues/8935 ) I also don't see any more.

Because I need to move on with this I may continue by trying to restore a backed up copy of my data into the newly created site to get rid of these issues. I will post back here if this fixes my problem. That said I think these issues should be left here as they occur as part of a normal upgrade so others may get them. 

I'm still happy to help debug the existing problems if you wish

**Comment by joolsr** [comment]
Hello

I tried a show processlist; when in mysqladmin and got this interesting output whilst it was running the background process delete_dynamic links

MariaDB [(none)]> show processlist;
+--------+------------------+-----------+------------------+---------+------+----------+-----------------------------------------------------------------------------------+----------+
| Id     | User             | Host      | db               | Command |
Time | State    |
Info                                                                             
| Progress |
+--------+------------------+-----------+------------------+---------+------+----------+-----------------------------------------------------------------------------------+----------+
| 182083 | root             | localhost | NULL             | Query  
|    0 | init     | show
processlist                                                                 
|    0.000 |
| 182112 | 1bd3e0294da19198 | localhost | 1bd3e0294da19198 | Query   |  
38 | updating | delete from tabVersion where ref_doctype='Communication'
and docname='b4bee73bf8' |    0.000 |
+--------+------------------+-----------+------------------+---------+------+----------+-----------------------------------------------------------------------------------+----------+
2 rows in set (0.00 sec)

ie the problematic query is this one 'delete from tabVersion where ref_doctype='Communication'
and docname='b4bee73bf8' ' which ran for more than 38s before disappearing.

This happens repeatedly for every email I delete ..

**Comment by tundebabzy** [comment]
Hello @joolsr
How many records do you have in the table?

**Comment by tundebabzy** [comment]
@joolsr 
Any updates regarding this issue?

**Comment by joolsr** [comment]
Well the specific query appears to be this one :-1: 

delete from tabVersion where ref_doctype='Communication'
and docname='b4bee73bf8'

I've decided to rebuild by server and export doctype info and import it into my new server. With the new server I have not seen any issues with high cpu usage and this query appearing. I think that the reason it appeared in the first place was a commit https://github.com/frappe/frappe/commit/ebbcd4593e3cb7fac06f2b3ad9c05fc3f4cf88dc which may then have been reverted or changed but has not ended up with the same query as just reinstalling completely afresh. 

So for this reason I would close this issue, unless anyone else has the same issue. There are certainly other issues that still need dealing with .

Many thanks for your help so far ..
